service/src/main/kotlin/fi/espoo/evaka/shared/domain/OperationalDays.kt

```diff
@@ -1,66 +1,109 @@
-// SPDX-FileCopyrightText: 2017-2022 City of Espoo
+// SPDX-FileCopyrightText: 2017-2024 City of Espoo
 //
 // SPDX-License-Identifier: LGPL-2.1-or-later
 
 package fi.espoo.evaka.shared.domain
 
+import fi.espoo.evaka.shared.ChildId
 import fi.espoo.evaka.shared.DaycareId
+import fi.espoo.evaka.shared.data.DateMap
+import fi.espoo.evaka.shared.data.DateSet
 import fi.espoo.evaka.shared.db.Database
-import java.time.DayOfWeek
 import java.time.LocalDate
-import java.time.Month
 
-fun LocalDate.isOperationalDate(operationalDays: Set<DayOfWeek>, holidays: Set<LocalDate>) =
-    operationalDays.contains(dayOfWeek) &&
-        // Units that are operational every day of the week are also operational during holidays
-        (operationalDays.size == 7 || !holidays.contains(this))
-
-data class OperationalDays(
-    val fullMonth: List<LocalDate>,
-    val generalCase: List<LocalDate>,
-    private val specialCases: Map<DaycareId, List<LocalDate>>
-) {
-    fun forUnit(id: DaycareId): List<LocalDate> = specialCases[id] ?: generalCase
-}
+fun Database.Read.getHolidays(range: FiniteDateRange): Set<LocalDate> =
+    createQuery {
+            sql("SELECT date FROM holiday WHERE between_start_and_end(${bind(range)}, date)")
+        }
+        .toSet<LocalDate>()
 
-fun Database.Read.operationalDays(year: Int, month: Month): OperationalDays {
-    val range = FiniteDateRange.ofMonth(year, month)
-    return operationalDays(range)
-}
+fun Database.Read.getOperationalDatesForChildren(
+    range: FiniteDateRange,
+    children: Set<ChildId>
+): Map<ChildId, Set<LocalDate>> {
+    data class PlacementRange(
+        val range: FiniteDateRange,
+        val childId: ChildId,
+        val unitId: DaycareId
+    )
+    val placements: Map<ChildId, DateMap<DaycareId>> =
+        createQuery {
+                sql(
+                    """
+        SELECT daterange(pl.start_date, pl.end_date, '[]') as range, pl.child_id, pl.unit_id
+        FROM placement pl
```

Comment:
Since this is for a range and not for a single day `realized_placement_one` couldn't be used. Instead fetched backup units into a separate DateMap and then merged the two.